x/mongo/driver/operation/find_and_modify.go

```diff
@@ -74,23 +72,23 @@ func buildFindAndModifyResult(response bsoncore.Document, srvr driver.Server) (F
 	famr := FindAndModifyResult{}
 	for _, element := range elements {
 		switch element.Key() {
-
 		case "value":
 			var ok bool
 			famr.Value, ok = element.Value().DocumentOK()
-			if !ok {
-				err = fmt.Errorf("response field 'value' is type document, but received BSON type %s", element.Value().Type)
+
+			// The 'value' field returned by a FindAndModify can be null in the case that no document was found.
```

Comment:
Calling attention to this bug-fix. There are a number of tests that have `FindAndModify` operations (`FindAndDelete` and `FindAndReplace` in particular; here's [an example](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/collection_test.go#L1236)) that do not find a matching document. In that case, the `value` returned by the server will be BSON null and `famr.Value` here will be `nil`. This situation _should_ result in `mongo.ErrNoDocuments`, and now that we're correctly returning errors from `build[type]Result` functions, this `if` check needs to also accept `value`s that are BSON null.

Reply:
The behavior seems sound, but the condition below is confusing. `DocumentOK()` returns either `return nil, false` or `return doc, true`, so the condition:
```go
if famr.Value != nil && !ok
```
will never evaluate to true.

Is there something else you can check before trying to read it as a `Document`, like if the value is empty?

Reply:
Excellent point. I've switched the `if` check to be `if element.Value().Type != bsontype.Null && !ok`. That should correctly return an error if the `value` field is neither BSON null _nor_ a BSON document.

Reply:
Makes sense 👍